src/main/java/dan200/computercraft/shared/peripheral/generic/data/DetailProviders.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * This file is part of ComputerCraft - http://www.computercraft.info
+ * Copyright Daniel Ratcliffe, 2011-2022. Do not distribute without permission.
+ * Send enquiries to dratcliffe@gmail.com
+ */
+package dan200.computercraft.shared.peripheral.generic.data;
+
+import dan200.computercraft.api.detail.BlockReference;
+import dan200.computercraft.api.detail.IDetailProvider;
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.fluids.FluidStack;
+
+import java.util.*;
+
+public final class DetailProviders
+{
+    private static final Map<Class<?>, Collection<IDetailProvider<?>>> allProviders = new HashMap<>();
+
+    public static synchronized <T> void registerProvider( Class<T> type, IDetailProvider<T> provider )
+    {
+        Objects.requireNonNull( type, "type cannot be null" );
+        Objects.requireNonNull( provider, "provider cannot be null" );
+
+        if ( !type.isAssignableFrom( BlockReference.class ) && !type.isAssignableFrom( ItemStack.class )
+            && !type.isAssignableFrom( FluidStack.class ) )
```

Comment:
Possibly worth using `==` here instead. I realise all three classes are final, but expresses the intent a little better.